### ISSUE TITLE
Client Delivery Limitations Mac OSX not working

### ISSUE THREAD

**Issue by coopersita** [revive-adserver/revive-adserver#597]
I have a contract ad with the following settings:

![banner-limits](https://cloud.githubusercontent.com/assets/636647/8690738/ee6084de-2a7d-11e5-85e7-c6750b7554d4.jpg)

The banner doesn't show up, even though I am using a Mac with 10.10.4 (Yosemite). If I remove the OS limitation, the banner shows up as expected.

I think it may an issue with PHPSniff. I went to their demo page, and it says my OS is Unknown.

I found this: http://sourceforge.net/p/phpsniff/patches/5/

"phpSniff does not correctly identify the new Intel-based macs which have a different identification in the HTTP_USER_AGENT system variable.
In the file phpSniff.core.php...
Update the regex string on line 243 from:
$regex_mac = '/(68[k0]{1,3})|(ppc mac os x)|([p\S]{1,5}pc)|(darwin)/i';
To:
$regex_mac = '/(68[k0]{1,3})|(Macintosh)|(ppc mac os x)|([p\S]{1,5}pc)|(darwin)/i';
Adding the full string option for "(Macintosh)" correctly identifies Intel-based Macs."

**Comment by erikgeurts** [comment]
@coopersita Thanks for your detailed report. Would you be able to fork, modify the code, and then file a pull request?

**Comment by mikelolasagasti** [comment]
maybe it will be more interesting moving from phpsniff to something more updated like https://github.com/gabrielbull/php-browser

**Comment by coopersita** [comment]
I think @mikelolasagasti is right. I tried implementing the patch, and it still didn't work (although I'm new to Revive, so I may not have done it in the right place).

In the forums I saw a lot of requests to be able to target mobile, so a more modern browser sniffer may be the solution.

**Comment by mbeccati** [comment]
Replacing phpSniff has been in my most-wanted list since a long time. We've started looking into replacements though.

**Comment by ianvm** [comment]
I agree, looking at the current code, phpSniff takes most of the CPU time away per request.

**Comment by mbeccati** [comment]
@mikelolasagasti thanks for the suggestion.

I was leaning towards ua-parser-php, but some early benchmarking was pretty disappointing, as the speed was 10 times slower than phpSniff. On the other hand gabrielbull/php-browser seems to be 15x faster than the old phpSniff!

**Comment by mikelolasagasti** [comment]
@mbeccati I can try to create a patch to migrate from phpSniff to php-browser if you're not doing it

**Comment by mbeccati** [comment]
We were thinking about restructuring the browser/os delivery limitations as well and in order to do so we'd ask for feedback on the forums, but I haven't had time just yet to write such post...

**Comment by rhapsodyv** [comment]
@mbeccati What do you are planning? We are about to work in some plugins/changes to keep track of the impressions delivered by restriction and bring to revive concepts like adunit and placements (as in dfp). We can unite efforts. What do you think?

**Comment by mbeccati** [comment]
@rhapsodyv No actual plans, other than replacing phpsniff and some (minor?) restructuring of browser/os limitations: the current browser/os lists come from phpSniff and are probably close to useless today.

**Comment by 2ge** [comment]
any update on this ?

**Comment by mbeccati** [comment]
@2ge I'm afraid not yet. It's still in my to-do list

**Comment by pixelchutes** [comment]
It might not be an exact drop-in replacement, but I find [Mobile_Detect](https://github.com/serbanghita/Mobile-Detect) to be a great library, with an easy to use API for UA matching, etc.

May not work specifically for Revive's needs, but it's a very active project with a decent sized audience.

*Crude Example:*
```php
if( $detect->isiOS() ) {
    $platform = 'ios';
} elseif( $detect->is('webOS') ) {
    $platform = 'webos';
} elseif( $detect->match('linux') ) {
    $platform = 'linux';
    if( $detect->isAndroidOS() ) {
        $platform = 'android';
    }
} elseif( $detect->match('windows') ) {
    $platform = 'win';
} elseif( $detect->match('macintosh|mac os x') ) {
    $platform = 'mac';
} else {
    $platform = 'other';
}
```

**Comment by mbeccati** [comment]
Fixed in 4.0.0
